### DISCUSSION TITLE
DNS 란

### DISCUSSION THREAD

**Discussion by michoball** [wanted-9th-3team/frontend-study#57] [프론트 기술 질문 공간]
DNS가 무엇인지 설명하고 사용자가 특정 도메인 (예를들어 google.com)에 접근하려할 때 DNS에서 어떻게 해당 도메인의 ip를 찾는지 알려주세요

**Comment by WooGie911** [comment]
DNS(Domain Name Syetem)란 사람이 읽을 수 있는 도메인 이름(예: www.google.com)을 머신이 읽을 수 있는 IP 주소(예: 173.194.126.240)로 변환하는 과정, 전체 시스템을 말합니다.

웹사이트의 데이터가 저장되어 있는 호스팅 서버는 인터넷 회선이 연결된 컴퓨터/장치인지라 IP 주소가 할당되어 있고 이 주소가 실제 웹사이트 주소라 할 수 있습니다. 
DNS 서버는 이런 IP 주소를 특정 도메인 주소와 같다는 기록을 저장해두고, 인터넷 사용자들이 도메인 주소를 검색했을 때 IP 주소로 연결되도록 해줍니다.

접근법에 대해 간단하게 설명하자면 브라우저에 google.com을 검색했다면, 먼저 DNS 서버로 도메인 주소가 전달됩니다. 
그리고 서버 내부에서 도메인 주소를 토대로 “google.com = 173.194.126.240”이라는 항목을 찾아내고 다시 브라우저에게 173.194.126.240의 IP 주소를 갖고 있는 호스팅 서버(해당 웹사이트 데이터가 저장된 곳)으로 가라고 지시를 내립니다. 
그러면 브라우저가 다시 IP 주소로 접속해서 웹사이트가 보이게 됩니다.

**Comment by michoball** [comment]
DNS는 도메인 네임 시스템의 약자로 특정 ip 에 사용자가 접근하고자 할 때 ip 주소는 쉽게 이해하기 어려운 구조로 되어있으므로 보다 쉽게 접근할 수 있도록 해주는 도메인이 모여있는 곳입니다. 

처음 유저가 특정 도메인주소를 주소창에 입력하면 해당 도메인에 매칭되는 ip주소를 local이나 root DNS 서버에 있는지 찾아봅니다. 이때 도메인을 찾을때는 주소의 맨 뒷부분 .com과 같은 부분부터 찾아 들어가서 해당 도메인을 찾습니다. 

도메인을 찾으면 해당 도메인에 매칭되는 ip주소를 가지는 서버에 데이터 요청을 보내게되고 사용자의 ip로 데이터를 전송하게 됩니다.

**Comment by javascriptbiggosoo** [comment]
DNS는 기본적으로 인터넷의 전화번호부라고 볼 수 있습니다. DNS는 웹 사이트의 도메인 이름 (예: [www.google.com)을](http://www.google.xn--com%29-8040a/) 해당 사이트의 IP 주소 (예: 172.217.0.4)로 변환하는 역할을 합니다. 이렇게 하면 컴퓨터는 웹 사이트에 연결할 수 있습니다. 이 과정은 DNS 조회(DNS lookup)라고 합니다.
DNS에서 IP 주소를 조회하는 과정은 다음과 같습니다:

사용자가 웹 브라우저에 도메인 이름(예: [www.example.com)을](http://www.example.xn--com%29-8040a/) 입력하면, 웹 브라우저는 이 도메인의 IP 주소를 알아내기 위해 DNS 조회를 요청합니다.
만약 로컬 DNS 서버가 해당 도메인의 IP 주소를 알고 있다면, 이 정보를 웹 브라우저에게 바로 반환합니다. 그렇지 않다면, 로컬 DNS 서버는 이 정보를 찾기 위해 인터넷의 다른 DNS 서버들에게 조회를 요청합니다.
이 요청은 루트 DNS 서버로 전달되고, 루트 DNS 서버는 TLD (Top Level Domain, 최상위 도메인) 서버를 가리키는 정보를 반환합니다. 이 TLD 서버는 도메인의 마지막 부분 (.com, .org 등)에 대한 정보를 가지고 있습니다.
TLD 서버는 해당 도메인의 authoritative DNS 서버를 가리키는 정보를 반환합니다.
여기에 ip가 포함되어 있습니다.

여기서 우리는 TLD가 항상 켜져있지 않다면 DNS를 이용할 수 없음을 짐작할 수 있습니다.
TLD 서버는 TLD를 관리하는 레지스트리에 의해 운영되며, 이들 서버는 항상 가동되어야 합니다. 이를 보증하기 위해, 레지스트리는 보통 다수의 TLD 서버를 전 세계 여러 위치에 분산 배치합니다. 이렇게 함으로써, 하나의 서버에 문제가 발생하더라도 다른 서버들이 도메인 이름 조회를 계속 처리할 수 있게 됩니다. 이러한 분산 시스템은 인터넷의 핵심 구성 요소의 가용성을 보증하며, 단일 지점 장애(Single Point of Failure)를 방지합니다.
그 외에도, 레지스트리는 서버 장비의 정기적인 유지보수와 업그레이드, 시스템의 보안을 유지하기 위한 다양한 조치, 그리고 서버에 부하가 너무 커지지 않도록 트래픽을 관리하는 등의 작업을 수행합니다. 이 모든 것들이 결합되어 TLD 서버가 항상 켜져 있음을 보증합니다.
